src/mango/src/mango_cursor.erl

```diff
@@ -44,27 +44,286 @@
+                    _ -> null
+                end,
+            Reasons = [{[{name, hd(Reason)}]}],
+            Analysis =
+                {[
+                    {usable, Usable},
+                    {reasons, Reasons},
+                    {ranking, Ranking},
+                    {covering, Covering}
+                ]},
+            Entry =
+                {[
+                    {index, mango_idx:to_json(Index)},
+                    {analysis, Analysis}
+                ]},
+            [Entry | Acc]
+        end,
+    lists:reverse(maps:fold(ToList, [], Candidates)).
+
+-type selector_hint() :: {indexable_fields, [field()]} | {unindexable_fields, [field()]}.
+-type selector_hints() :: {[selector_hint()]}.
+
+-spec extract_selector_hints(selector()) -> selector_hints().
+extract_selector_hints(Selector) ->
+    DreyfusAvailable = dreyfus:available(),
+    NouveauEnabled = nouveau:enabled(),
+    AsIndex =
+        fun(Module) ->
+            case Module of
+                mango_cursor_view -> [{mango_idx_view, json}];
+                mango_cursor_text when DreyfusAvailable -> [{mango_idx_text, text}];
+                mango_cursor_nouveau when NouveauEnabled -> [{mango_idx_nouveau, nouveau}];
+                _ -> []
+            end
+        end,
+    Modules = lists:flatmap(AsIndex, ?CURSOR_MODULES),
+    Populate =
+        fun({Module, IndexType}) ->
+            AllFields = sets:from_list(mango_selector:fields(Selector)),
+            Normalize = fun(N) -> hd(string:split(N, ":")) end,
+            IndexableFields = sets:from_list(
+                lists:map(Normalize, Module:indexable_fields(Selector))
+            ),
+            UnindexableFields = sets:subtract(AllFields, IndexableFields),
+            {[
+                {type, IndexType},
+                {indexable_fields, sets:to_list(IndexableFields)},
+                {unindexable_fields, sets:to_list(UnindexableFields)}
+            ]}
+        end,
+    lists:map(Populate, Modules).
+
+-type explain_attribute() ::
+    {dbname, database()}
+    | {index, ejson()}
+    | {partitioned, boolean()}
+    | {selector, ejson()}
+    | {opts, ejson()}
+    | {limit, integer()}
+    | {skip, integer()}
+    | {fields, fields()}
+    | {index_candidates, [candidate_index()]}
+    | {selector_hints, selector_hints()}
+    | {atom(), any()}.
+-type explain_response() :: {[explain_attribute()]}.
+
+-spec explain(#cursor{}) -> explain_response().
 explain(#cursor{} = Cursor) ->
     #cursor{
-        index = Idx,
+        db = Db,
+        index = Index,
         selector = Selector,
         opts = Opts0,
         limit = Limit,
         skip = Skip,
         fields = Fields
     } = Cursor,
-    Mod = mango_idx:cursor_mod(Idx),
+    {ModExplain, DbName, JSON, Partitioned} =
+        case Index of
+            none ->
+                {
+                    [],
+                    couch_db:name(Db),
+                    null,
+                    null
```

Comment:
What are the possible `mango_idx:partioned(Idx)` return values? Would `null` be a new return type for it? 

I looked in `mango_idx(Idx)` but go confused as I found a mix of `false/true/undefined/db_default` values.

Reply:
The original intention about using `null` here is to reflect JSON-wise that this property is not applicable.  There is no index in this case so it is not either global or partitioned.  Similarly, it does not have a definition.  As far as I know, `jiffy` can only create `null` if the `null` atom is used.

Reply:
@willholley what do you think?

Reply:
I don't think there is such thing as "no index" so I think this can never be null? `_all_docs` is a primary index and can be partitioned or global. If a partitioned flag is not present, that indicates a global index.

Reply:
ah if we're emitting json then `null` would work.

Reply:
Later we discussed this with @willholley off-line.  He asked me to decouple the `partitioned` field from the index.  That is the property of the query itself so it should say either `true` of `false` depending on the URL, independently from the fact that if there was a suitable index or not.